### PR TITLE
[#7] [#12] [Infrastructure] Setup AWS ALB [Infrastructure] Setup AWS S3

### LINKED ISSUE #7: [Infrastructure] Setup AWS ALB
## Why

AWS Application Load Balancer (ALB) is a load balancing service that distributes incoming traffic across multiple targets, such as EC2 instances, containers, and IP addresses, in one or more Availability Zones. By setting up an ALB for your application, you can improve the availability and fault tolerance of your application by distributing incoming traffic across multiple targets. This can help ensure that your application can handle a high volume of traffic and continue to operate even if one or more of your targets fail.

Additionally, using multiple Availability Zones can further improve the availability and fault tolerance of your application by ensuring that it can continue to operate even if an entire Availability Zone becomes unavailable.

## Acceptance Criteria

 - The AWS ALB service has been set up for the application.
 - The ALB is configured to operate in multiple Availability Zones.
 - The ALB is properly routing traffic to the correct targets.

### LINKED ISSUE #12: [Infrastructure] Setup AWS S3
## Why

AWS Simple Storage Service (S3) is an object storage service that offers industry-leading scalability, availability, and security for storing and retrieving data objects. By setting up S3 for our application, we can store and retrieve our application data objects with high performance and reliability. This can help ensure that our application can handle a high volume of data and continue to operate even if one or more of our data objects become unavailable.

## Acceptance Criteria

 - The AWS S3 service has been set up for the application.
 - An S3 bucket has been created and is properly configured for the application.
 - The application data objects are being properly stored in the S3 bucket.
 - The application data objects can be accessed and retrieved from the S3 bucket.

### PR DESCRIPTION
**Pull request by Nihisil**
- Close #7 
- Close #12

## What happened 👀

Set up an Application Load Balancer (ALB) and an Amazon Simple Storage Service (S3) bucket for logging ALB activity. 

The S3 service can be leveraged in the future for storing application content files. 

Since ALB does not provide the capability to log to Amazon CloudWatch, an S3 bucket is required: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb#access_logs

## Proof Of Work 📹

Plan was applied successfully (on test env):

image

`$ terraform plan` (command output below, it executed planning on TF cloud)

```
Terraform will perform the following actions:

  # module.alb.aws_lb.main will be created
  + resource "aws_lb" "main" {
      + arn                        = (known after apply)
      + arn_suffix                 = (known after apply)
      + desync_mitigation_mode     = "defensive"
      + dns_name                   = (known after apply)
      + drop_invalid_header_fields = true
      + enable_deletion_protection = false
      + enable_http2               = true
      + enable_waf_fail_open       = false
      + id                         = (known after apply)
      + idle_timeout               = 60
      + internal                   = false
      + ip_address_type            = (known after apply)
      + load_balancer_type         = "application"
      + name                       = "alex-ic-staging-alb"
      + preserve_host_header       = false
      + security_groups            = (known after apply)
      + subnets                    = [
          + "subnet-009c63829ea5098d6",
          + "subnet-032fa118f45ec3cef",
          + "subnet-09073722ed8d77b04",
        ]
      + tags_all                   = {
          + "Environment" = "staging"
          + "Owner"       = "alex-ic"
        }
      + vpc_id                     = (known after apply)
      + zone_id                    = (known after apply)

      + access_logs {
          + bucket  = "alex-ic-staging-alb-log"
          + enabled = true
        }

      + subnet_mapping {
          + allocation_id        = (known after apply)
          + ipv6_address         = (known after apply)
          + outpost_id           = (known after apply)
          + private_ipv4_address = (known after apply)
          + subnet_id            = (known after apply)
        }
    }

  # module.alb.aws_lb_listener.app_http will be created
  + resource "aws_lb_listener" "app_http" {
      + arn               = (known after apply)
      + id                = (known after apply)
      + load_balancer_arn = (known after apply)
      + port              = 80
      + protocol          = "HTTP"
      + ssl_policy        = (known after apply)
      + tags_all          = {
          + "Environment" = "staging"
          + "Owner"       = "alex-ic"
        }

      + default_action {
          + order            = (known after apply)
          + target_group_arn = (known after apply)
          + type             = "forward"
        }
    }

  # module.alb.aws_lb_target_group.target_group will be created
  + resource "aws_lb_target_group" "target_group" {
      + arn                                = (known after apply)
      + arn_suffix                         = (known after apply)
      + connection_termination             = false
      + deregistration_delay               = "100"
      + id                                 = (known after apply)
      + ip_address_type                    = (known after apply)
      + lambda_multi_value_headers_enabled = false
      + load_balancing_algorithm_type      = (known after apply)
      + name                               = (known after apply)
      + port                               = 4000
      + preserve_client_ip                 = (known after apply)
      + protocol                           = "HTTP"
      + protocol_version                   = (known after apply)
      + proxy_protocol_v2                  = false
      + slow_start                         = 0
      + tags_all                           = {
          + "Environment" = "staging"
          + "Owner"       = "alex-ic"
        }
      + target_type                        = "ip"
      + vpc_id                             = "vpc-031c6a72019833218"

      + health_check {
          + enabled             = true
          + healthy_threshold   = 3
          + interval            = 5
          + matcher             = "200-299"
          + path                = "/_health/readiness"
          + port                = "4000"
          + protocol            = "HTTP"
          + timeout             = 3
          + unhealthy_threshold = 2
        }

      + stickiness {
          + cookie_duration = 86400
          + enabled         = true
          + type            = "lb_cookie"
        }

      + target_failover {
          + on_deregistration = (known after apply)
          + on_unhealthy      = (known after apply)
        }
    }

  # module.s3.aws_s3_bucket.alb_log will be created
  + resource "aws_s3_bucket" "alb_log" {
      + acceleration_status         = (known after apply)
      + acl                         = (known after apply)
      + arn                         = (known after apply)
      + bucket                      = "alex-ic-staging-alb-log"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = true
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + object_lock_enabled         = (known after apply)
      + policy                      = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = {
          + "Environment" = "staging"
          + "Owner"       = "alex-ic"
        }
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + cors_rule {
          + allowed_headers = (known after apply)
          + allowed_methods = (known after apply)
          + allowed_origins = (known after apply)
          + expose_headers  = (known after apply)
          + max_age_seconds = (known after apply)
        }

      + grant {
          + id          = (known after apply)
          + permissions = (known after apply)
          + type        = (known after apply)
          + uri         = (known after apply)
        }

      + lifecycle_rule {
          + abort_incomplete_multipart_upload_days = (known after apply)
          + enabled                                = (known after apply)
          + id                                     = (known after apply)
          + prefix                                 = (known after apply)
          + tags                                   = (known after apply)

          + expiration {
              + date                         = (known after apply)
              + days                         = (known after apply)
              + expired_object_delete_marker = (known after apply)
            }

          + noncurrent_version_expiration {
              + days = (known after apply)
            }

          + noncurrent_version_transition {
              + days          = (known after apply)
              + storage_class = (known after apply)
            }

          + transition {
              + date          = (known after apply)
              + days          = (known after apply)
              + storage_class = (known after apply)
            }
        }

      + logging {
          + target_bucket = (known after apply)
          + target_prefix = (known after apply)
        }

      + object_lock_configuration {
          + object_lock_enabled = (known after apply)

          + rule {
              + default_retention {
                  + days  = (known after apply)
                  + mode  = (known after apply)
                  + years = (known after apply)
                }
            }
        }

      + replication_configuration {
          + role = (known after apply)

          + rules {
              + delete_marker_replication_status = (known after apply)
              + id                               = (known after apply)
              + prefix                           = (known after apply)
              + priority                         = (known after apply)
              + status                           = (known after apply)

              + destination {
                  + account_id         = (known after apply)
                  + bucket             = (known after apply)
                  + replica_kms_key_id = (known after apply)
                  + storage_class      = (known after apply)

                  + access_control_translation {
                      + owner = (known after apply)
                    }

                  + metrics {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }

                  + replication_time {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }
                }

              + filter {
                  + prefix = (known after apply)
                  + tags   = (known after apply)
                }

              + source_selection_criteria {
                  + sse_kms_encrypted_objects {
                      + enabled = (known after apply)
                    }
                }
            }
        }

      + server_side_encryption_configuration {
          + rule {
              + bucket_key_enabled = (known after apply)

              + apply_server_side_encryption_by_default {
                  + kms_master_key_id = (known after apply)
                  + sse_algorithm     = (known after apply)
                }
            }
        }

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }

      + website {
          + error_document           = (known after apply)
          + index_document           = (known after apply)
          + redirect_all_requests_to = (known after apply)
          + routing_rules            = (known after apply)
        }
    }

  # module.s3.aws_s3_bucket_acl.alb_log_bucket_acl will be created
  + resource "aws_s3_bucket_acl" "alb_log_bucket_acl" {
      + acl    = "private"
      + bucket = (known after apply)
      + id     = (known after apply)

      + access_control_policy {
          + grant {
              + permission = (known after apply)

              + grantee {
                  + display_name  = (known after apply)
                  + email_address = (known after apply)
                  + id            = (known after apply)
                  + type          = (known after apply)
                  + uri           = (known after apply)
                }
            }

          + owner {
              + display_name = (known after apply)
              + id           = (known after apply)
            }
        }
    }

  # module.s3.aws_s3_bucket_policy.allow_elb_logging will be created
  + resource "aws_s3_bucket_policy" "allow_elb_logging" {
      + bucket = (known after apply)
      + id     = (known after apply)
      + policy = (known after apply)
    }

  # module.s3.aws_s3_bucket_public_access_block.alb_log will be created
  + resource "aws_s3_bucket_public_access_block" "alb_log" {
      + block_public_acls       = true
      + block_public_policy     = true
      + bucket                  = (known after apply)
      + id                      = (known after apply)
      + ignore_public_acls      = true
      + restrict_public_buckets = true
    }

  # module.security_group.aws_security_group.alb will be created
  + resource "aws_security_group" "alb" {
      + arn                    = (known after apply)
      + description            = "ALB Security Group"
      + egress                 = (known after apply)
      + id                     = (known after apply)
      + ingress                = (known after apply)
      + name                   = "alex-ic-staging-alb-sg"
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags                   = {
          + "Name" = "alex-ic-staging-alb-sg"
        }
      + tags_all               = {
          + "Environment" = "staging"
          + "Name"        = "alex-ic-staging-alb-sg"
          + "Owner"       = "alex-ic"
        }
      + vpc_id                 = "vpc-031c6a72019833218"
    }

  # module.security_group.aws_security_group_rule.alb_egress will be created
  + resource "aws_security_group_rule" "alb_egress" {
      + cidr_blocks              = [
          + "0.0.0.0/0",
        ]
      + description              = "From ALB to app"
      + from_port                = 4000
      + id                       = (known after apply)
      + protocol                 = "tcp"
      + security_group_id        = (known after apply)
      + security_group_rule_id   = (known after apply)
      + self                     = false
      + source_security_group_id = (known after apply)
      + to_port                  = 4000
      + type                     = "egress"
    }

  # module.security_group.aws_security_group_rule.alb_ingress_http will be created
  + resource "aws_security_group_rule" "alb_ingress_http" {
      + cidr_blocks              = [
          + "0.0.0.0/0",
        ]
      + description              = "From HTTP to ALB"
      + from_port                = 80
      + id                       = (known after apply)
      + protocol                 = "tcp"
      + security_group_id        = (known after apply)
      + security_group_rule_id   = (known after apply)
      + self                     = false
      + source_security_group_id = (known after apply)
      + to_port                  = 80
      + type                     = "ingress"
    }

Plan: 10 to add, 0 to change, 0 to destroy.
```

### CODE REVIEW

File: modules/security_group/main.tf [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#tfsec:ignore:aws-ec2-no-public-ingress-sgr
+resource "aws_security_group_rule" "alb_ingress_http" {
+  type              = "ingress"
+  security_group_id = aws_security_group.alb.id
+  protocol          = "tcp"
+  from_port         = 80
+  to_port           = 80
+  cidr_blocks       = ["0.0.0.0/0"]
+  description       = "From HTTP to ALB"
+}
```
Comment (longnd) [inline modules/security_group/main.tf:20]: how about:
- ALB rules to redirect HTTP connection -> HTTPS
- security group rules allow HTTPS traffic only (port 443)

example: https://github.com/nimblehq/worpt-infrastructure/blob/c4e8f8f3641a095d09c683484489e5caf450c6fb/modules/alb/main.tf#L41-L55

Comment (andyduong1920) [reply modules/security_group/main.tf:20]: For IC, do we really need SSL @longnd :D, if the SSL is required, then the developer need to purchase a domain as well :D

**Review by andyduong1920 — APPROVED** [review]

File: modules/security_group/main.tf (hunk 1 above)
Comment (Nihisil) [reply modules/security_group/main.tf:20]: Yes, we cannot utilize HTTPS with the domain that ALB provided; we need to attach the domain and certificate to AWS if we wish to use HTTPS here.

https://stackoverflow.com/questions/68567684/why-does-awss-application-load-balancer-require-a-custom-domain-when-using-http

@longnd What do you think?

Comment (longnd) [reply modules/security_group/main.tf:20]: yes, you are right. I did not think/forget that we don't have a domain for used in the IC, thus can't issue a certificate for use by the ALB. I will resolve this comment 🙏

**Review by longnd — APPROVED** [review]

**Pull request merged**